arbos/programs/native.go

```diff
@@ -233,11 +231,27 @@ func activateProgramInternal(
 	return info, asmMap, err
 }
 
+// getLocalAsm gets asm for local target and recompiles it if not found.
+// Existence of asms for other configured targets is checked and also recompiled if not found (other targets are needed for multi-target recording).
 func getLocalAsm(statedb vm.StateDB, moduleHash common.Hash, addressForLogging common.Address, code []byte, codehash common.Hash, maxWasmSize uint32, pagelimit uint16, time uint64, debugMode bool, program Program, runCtx *core.MessageRunContext) ([]byte, error) {
 	localTarget := rawdb.LocalTarget()
-	localAsm, err := statedb.TryGetActivatedAsm(localTarget, moduleHash)
-	if err == nil && len(localAsm) > 0 {
-		return localAsm, nil
+	targets := runCtx.WasmTargets()
+	// even though we need only asm for local target, make sure that all configured targets are available as they are needed during multi-target recording of a program call
+	asmMap, missingTargets, err := statedb.ActivatedAsmMap(targets, moduleHash)
```

Comment:
only in RecordingMode we should check and recompile other targets then local target

Reply:
and for CommitMode